src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -612,8 +615,57 @@ public virtual ProtoCore.Type GetTypeHintForOutput(int index)
         {
              return ProtoCore.TypeSystem.BuildPrimitiveTypeObject(ProtoCore.PrimitiveType.kTypeVar);
         }
-        #endregion
 
+      
+        /// <summary>
+        /// Gets or sets a value indicating whether the node is frozen.
+        /// If the node is set to freeze by the user, then Freeze always true.
+        /// Otherwise, if any of the node has input node frozen, then Freeze is true.        
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this node is frozen; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsFrozen
+        {
+            get
+            {
+                var gatheredInputFrozenNodes = new List<NodeModel>();
+                CheckIfAnyUpstreamNodeIsFrozen(this, gatheredInputFrozenNodes);
+
+                if (gatheredInputFrozenNodes.Any())
+                    return true;
+ 
+                return explictFrozen;
+            }
+            set
+            {
+                explictFrozen = value;                 
+                RaisePropertyChanged("IsFrozen");
+                OnNodeModified();
+            }
+        }
+
+        private void CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, List<NodeModel> listInputNodes )
```

Comment:
See the note above on simplifying this method. We don't require the list of frozen nodes.